### ISSUE TITLE
RFC: Where to put official Kubernetes packages / templates / etc.?

### ISSUE THREAD

**Issue by zmerlynn** [kubernetes/kubernetes#17027]
We're starting work on building more baked, fully formed sets of resources and images running on Kubernetes (see #16517, which is still very much in pre-alpha, but defines an epoch to get Spark to a respectable place). Right now, we (the Kubernetes team) have been checking the pieces for this into the `examples/` directory into the main https://github.com/kubernetes/kubernetes repo. That's starting to feel wrong for a few reasons:

* The collection of resources and images for "apps" might use advanced features that only exist on `master`, but typically are designed to work back to old versions. If they *do* use advanced features, typically I would suggest templatizing it somehow (more on that in a bit).
* There's a fundamental difference between teaching examples, and packages. For instance, you might use `nginx` to teach Services, but an `nginx` package may actually look pretty different (it should include a `ConfigResource`, once it exists, and more metadata in general). Teaching examples don't have to be full-blown apps, and full-blown apps don't have to teach.

In addition, we've now got both [Deployment Manager](https://github.com/kubernetes/deployment-manager), [Helm](https://github.com/deis/helm) and [OpenShift Templates](https://docs.openshift.org/latest/dev_guide/templates.html) as interesting targets to write templates/packages for. It'd be nice to package this metadata in one place with the package. ~~(Unfortunately, helm (unlike DM) can't point at an alternate git path, so that one is actually tricky/possibly wasteful to consolidate under our repo as well. This is possibly a temporary limitation, though.)~~

**Current Proposal**:

* Create a https://github.com/kubernetes/application-dm-templates repo to host Deployment Manager templates, satisying https://github.com/kubernetes/deployment-manager/issues/82, splitting it from https://github.com/kubernetes/deployment-manager for all the reasons laid out in https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-157785896.

* Create a https://github.com/kubernetes/application-images repo to host image build content for containerized applications: Dockerfiles, startup scripts, config files, etc.

Both of these can be maintained by (separate) teams with fairly open memberships.

If there is interest in an additional repository to host Helm charts, we can create such a repo, but for the upstream https://github.com/helm/charts is the obvious spot. Similarly, if there is interest in hosting OpenShift templates, we can create such a repo, but the upstream https://github.com/openshift/origin/tree/master/examples is the obvious spot. We will link there in the k8s docs as well.

cc @brendandburns @bgrant0607 @aronchick @kubernetes/sig-big-data @jackgr

**Comment by bparees** [comment]
I like the idea of moving them out of the k8s repository.  In openshift we've taken the approach of creating standalone repositories for fully baked examples of things.  That way they can live and evolve on their own.  The downside is that it's harder to include them in tests to ensure they don't bit-rot.  

Here are a couple of them in case you want to see what they look like:
https://github.com/openshift/cakephp-ex
https://github.com/openshift/rails-ex
https://github.com/openshift/nodejs-ex
https://github.com/openshift/postgresql/tree/master/examples/replica
https://github.com/openshift/jenkins

etc

Having a separate repo for each example (rather than a single repo of examples) has some advantages when it comes to managing the issue list, and also since we git clone the repos during our build process (eg building a new rails application image), it's more efficient to have more fine grained repos.  It's also nicer for people who want to fork and work with a particular example.  

Of course it does have some more administrative overhead.

**Comment by zmerlynn** [comment]
Yeah, I was considering the "repo per package" approach as well, and I can't tell if it's worthwhile to go that file-grained or not. It certainly makes sense from an issue-subscription point-of-view, and means that a package contributor only has to subscribe to part of the firehose (which is a huge advantage), but it makes the top-level repo that much more balkanized. I guess that's not *that* big a deal, especially if they have a well-patterned name, like just strictly do `application/blah` or `example/blah` or `package/blah`

**Comment by timothysc** [comment]
fwiw repo-per-package enters into the distribution space, which is kinda where we're going right.?.?

e.g. Meaning 
- well defined semantics around versioning etc.   
- ability to earn package acceptance 
- feedback mechanics for failed packages 
- etc, etc.

**Comment by zmerlynn** [comment]
Fair. I might be sold.

**Comment by timothysc** [comment]
This is such a broad topic that will literally impact everyone, I'm all for adding this as a community topic.

**Comment by zmerlynn** [comment]
@timothysc: Seems reasonable. I'll see if I can get my microphone not to freak out over Zoom first. :p

**Comment by zmerlynn** [comment]
Actually, it looks like after https://github.com/deis/helm/pull/162, you can actually add alternate git remotes as chart repos in helm. I think certainly helm, there might be a downside in keeping one-package-per-repo. And thinking about it more, if you ever want to do anything that involves more than one package (e.g. maintain a distribution of packages like a Spark/Hadoop "package set" or something), you may actually want atomicity across the repo. So I'm a little torn now.

**Comment by jackgr** [comment]
With [dm](https://github.com/kubernetes/deployment-manager), you can put types in any Github repository. You just need a top level `/types` folder. Types live in version folders (e.g., types/redis/v1, types/replicatedservice/v2, etc.). See the [README](https://github.com/kubernetes/deployment-manager/blob/master/README.md) for details.

I'm not a fan of repo per type, since that would led to a LOT of repos, with all the overhead of setting up and maintaining each repo.

A "vetted" repo, as advocated by @zmerlynn seems a nice approach. To paraphrase his recommendation:

"Each type would require an owner, appropriate README, etc. A top-level README would describe type maintainer duties (and how often we might check - we need some sort of conformance testing there, too, which is an interesting burden)."

The folks I've been talking to consider the last part of that description - conformance testing - to be the most important. Every type definition should be supported by a test suite that runs automatically to verify its integrity.

Note that I've been talking about types, not packages. dm deals in types, not packages:

* Types are composable, and can be parameterized. 
* Packages make sense for nodes, and helm will install them on your local host. 
* Types make sense for containers, and dm will install them in your cluster.

See [this comparison](https://groups.google.com/forum/#!topic/kubernetes-sig-config/_sPQiBPUcc4) of dm and helm for details.

**Comment by bgrant0607** [comment]
Agree that examples don't belong in the main repo. This was discussed when moving to the kubernetes github org, but we need to change the way we test many of the examples to make that feasible. cc @mikedanese 

I'm fine with one repo per example, if we put them in a separate github org. They'd create too much noise in the main kubernetes repo.

Related: #15193, #5465, #4926

cc @jeffmendoza @amygdala

**Comment by davidopp** [comment]
It sounds like Zach's most recent comment was leaning towards one repo, which seems fine to me. IIUC the only downside is you'd be all-or-nothing in subscribing to the notification emails?

**Comment by davidopp** [comment]
Oh, I guess @timothysc also had some arguments for one repo per package... It does seem like a lot of repos, though. (Though assuming they're all sub-repos of one repo, it might not be too bad...)

**Comment by kargakis** [comment]
Separate org (@kubernetes-examples ?) and one repo per example seems appealing..

**Comment by feiskyer** [comment]
+1 for separate github org

**Comment by jimmidyson** [comment]
I also like separate org & repo per example.

**Comment by mattf** [comment]
what if examples are just starting points for developers wanting to work w/ kubernetes and they keep living in the kubernetes repo as they are today. packages/templates/etc are a step beyond examples, with dm / helm / openshift templates as proposals for how to develop and manage them.

examples would be...

 * maintained by the kubernetes community
 * functional with master
 * single version (e.g. no mariadbv5 & mariadbv10)
 * a demonstration of specific kubernetes capabilities or an enabling use case

...leaving everything else to dm / helm / openshift templates / others

"official" packages/templates/etc basically means picking a preferred manager from {dm, helm, openshift templates, other?} or forcing some sort of consolidation.

**counter proposal**: firm up what it means to be an example and let the package/template/etc managers grow on top of kubernetes organically. leave the option open to pick a preferred manager in the future.

@zmerlynn i think you're right that there's a real need here for kubernetes. it's one that helm, dm, openshift are exploring. imho, it's best to enable them (make sensible model changes for them) and let them explore for now.

**Comment by zmerlynn** [comment]
> what if examples are just starting points for developers wanting to work w/ kubernetes and they keep living in the kubernetes repo as they are today.

That was my original point. I'd actually like `examples/` to become more like ABC building block examples, and we should eliminate whole apps (unless the blocks build a single one up). We can use them to demonstrate, like you do with example config files that have *all* the options, etc. But `examples/` right now is just full of very similar looking apps that aren't actually revealing a whole ton, because it's in part getting conflated with an *apps* directory.

I think `examples/` really should just be limited to the examples that teach, too.

> counterproposal: firm up what it means to be an example and let the package/template/etc managers grow on top of kubernetes organically. leave the option open to pick a preferred manager in the future.

There's definitely a need, and we could actually consolidate some of them on our GitHub. I'm not suggesting that we eliminate competition, but I think it would be nice to have a place to put apps that were really well vetted and had support for multiple managers.

**Comment by zmerlynn** [comment]
@jackgr: Updated the top comment to reflect conformance more strongly. I don't necessarily want to get bogged down in it from the get go (I think we need an "if you build it, they will come", and in part that includes the test pieces), but I do think it's an important goal to have, and possibly a reason to centralize.

**Comment by jackgr** [comment]
@zmerlynn Thinking about your comment re: support for multiple managers.  Currently, the managers are quite different... 

* helm is YAML without parameters and runs client side only.
* openshift is flat parameterized YAML and runs client and server side.
* dm is parameterized YAML with types and runs client and server side.

Trying to span all 3 (or more) configuration formats for a single application doesn't sound very appealing. Likewise, putting applications based on different managers in the same repository doesn't sound very appealing. At best, you'd get a least common denominator effect, where clients couldn't take advantage of server side features, such as listing the types and instances deployed in a cluster, because helm doesn't support them.

More importantly, we have opportunity to move beyond configuration format discussions, and define a firmer notion of `application` for Kubernetes. While an application will likely be defined by one or more templates, it also needs metadata that tools and controllers can use to reason about it... e.g., metadata that:

* describes its monitoring, logging and alerting configuration, 
* allows it to be listed in a marketplace, 
* helps Ubernetes schedule it and/or its constituent components, 
* etc.

This higher order metadata may be hard to support across 3 (or more) different configuration formats.

**Comment by zmerlynn** [comment]
@jackgr: I think what you'd end up with is a `helm` package that was very limited, then DM/OS templates that were much richer, and probably it would depend on the package maintainer as to what was even picked up and used.

Do you have a better proposal? The existing `examples/` dir is not tenable, and one of the commonalities of all of those templates is a set of supported Docker images, regardless of the templating and types.

**Comment by jackgr** [comment]
No dispute with the desire to make /examples a collection of curated tutorials, since that's its purpose. Also, no dispute that collections of real world applications and application components (microservices?) are much needed, and that they should be distinct from /examples.

For now, to your point, since Docker images and Kubernetes YAML files are only de facto standardized foundations at this point in time, /examples should probably limit itself to those technologies.

Meanwhile, the community can and will, I'm sure, build type registries, package repositories, etc., that people can use for real world applications. Why don't we let that process run for a while, and see which technologies are embraced by the community for that purpose. When facto standards for templatization have emerged, and the community has gained experience with them, then /examples can start to include examples that leverage them.

**Comment by davidopp** [comment]
There's a limit to how much we can punt on picking (a) solution(s). The whole point of the project Zach is doing is to address the common (and valid) complaint we get that Kubernetes can't easily be used for production deployments of Spark, Cassandra, etc. There are a number of pieces of this -- missing k8s features (like nominal services), inadequate or nonexistant k8s configs for these systems, no package manager for users to install these apps, the apps are not distributed as Docker containers, etc.  To truly solve the problem we have to address all of these pieces, either ourselves or by using existing systems. But we have to give users at least one turnkey end-to-end solution. 

I think we need to set up a separate repo with these vetted, end-to-end, production-grade configurations, and not reuse examples/, for the reasons Zach mentioned. Whether we choose one packaging solution initially and then add more later, or try to do several up-front, is less important to me, but we should make sure to choose an approach that will be accepted by/popular with the community so that others will contribute to this effort.

To put it another way -- Zach is trying to find and execute the quickest path to publishing production deployments of Spark, Cassandra, etc., not solve the packaging/configuration/application description problem in general.

**Comment by jstrachan** [comment]
I like the idea of a federated set of github repos; so different communities can curate their own charts/packages without requiring approval from some central organisation for every change to a chart/package - this will help accelerate the creation of OOTB charts/packages and fast iteration by their users. 

Though it would be nice to have a global index of all chart/package git repos; so that folks can search across everything if they wish - like querying docker hub - but where each 'user' in the hub is really a separate git repo.

Maybe there could be multiple 'global indexes' which can curate or verify charts/packages for inclusion; e.g. running CI / CD processes before approving them or something. 

I can see folks also wanting curated subsets of packages which have been vetted to work nicely (e.g. with proper logging & metrics or whatever). On the [fabric8 project](http://fabric8.io/) for example we've been spending a lot of time iterating on [our kubernetes packages for DevOps and CI / CD](https://github.com/fabric8io/fabric8-devops) so that things like gogs, jenkins, nexus, artifactory, gerrit, sonarqube, hubot, slack, taiga, letschat, gitlab, kibana, grafana, prometheus, elasticsearch et al - all work seamlessly together OOTB.

TL;DR - separate organisations & repos sounds like a great idea; then folks can combine other repos into new repos too to make consolidated repos

**Comment by davidopp** [comment]
How about one new repo (and one organization), with sub-repos for each app/framework?

**Comment by zmerlynn** [comment]
> How about one new repo (and one organization), with sub-repos for each app/framework?

@davidopp: I'm trying to figure out what distinction you're making here. Are you proposing a common repo for metadata (e.g. a place for helm / DM / etc to grab), with separate repos for each thing? I considered that briefly, but the referential integrity makes me not want to do that: You really want to store k8s resources at the same time as you're updating a DM template or helm package, I think.

I'm about to move on and build a DM template, though, so maybe I'll get a better sense of all the moving pieces for the slice.

I might be misunderstanding what you're meaning, though.

**Comment by zmerlynn** [comment]
FWIW, I'm leaning towards two wildly different directions:

1. A central repo (and possibly separate org) where we are fairly open on who can contribute, though I expect you'd be expected to play within your playground. (I expect this would probably still be Google CLA, though.)
1. A loose confederacy of Git URLs, centrally maintained, where one can find DM, OpenShift, helm, etc. templates. At the limit, it would be nice if each of these engines just supported a native configuration file that could take a large number of these URLs, and we could just maintain a central copy somewhere.

One thing pushing me towards doing #1 anyways: These actually don't have to be exclusive. We can have #1 as e.g. "Kubernetes.io Applications", and have it as part of the federation. So even if we end up with a bunch of islands, we (the Kubernetes team) may want our own, unless/until it's clear the community has stepped up to do that job awesomely - and even then, we could take the best of breed and host them.

**Comment by timothysc** [comment]
As a community option 1 would have to be cncf owned, not directly under googles perview.

**Comment by zmerlynn** [comment]
@timothysc: Fixed comment. I expect at some point the @googlebot will magically transform into @cncfbot anyways.

**Comment by bacongobbler** [comment]
> Unfortunately, helm (unlike DM) can't point at an alternate git path, so that one is actually tricky/possibly wasteful to consolidate under our repo as well. This is possibly a temporary limitation, though.

Just poking in this thread now, but helm does have support for alternate git paths:

```
$ helm repo add bacongobbler git@github.com:bacongobbler/charts
---> Cloning into '/Users/bacongobbler/.helm/cache/bacongobbler'...
$ helm search redis
	bacongobbler/redis-cluster (redis-cluster 0.0.5) - Highly available Redis cluster with multiple sentinels and standbys.
	bacongobbler/redis-standalone (redis-standalone 0.0.1) - Standalone Redis Master
	redis-cluster (redis-cluster 0.0.5) - Highly available Redis cluster with multiple sentinels and standbys.
	redis-standalone (redis-standalone 0.0.1) - Standalone Redis Master
$ cat ~/.helm/config.yaml
repos:
  default: charts
  tables:
  - name: charts
    repo: https://github.com/deis/charts
  - name: bacongobbler
    repo: git@github.com:bacongobbler/charts
workspace: {}
```

If we change the `default` value in config.yaml to `bacongobbler`, it now becomes the default repository:

```
$ $EDITOR ~/.helm/config.yaml
$ helm search redis
	charts/redis-cluster (redis-cluster 0.0.5) - Highly available Redis cluster with multiple sentinels and standbys.
	charts/redis-standalone (redis-standalone 0.0.1) - Standalone Redis Master
	redis-cluster (redis-cluster 0.0.5) - Highly available Redis cluster with multiple sentinels and standbys.
	redis-standalone (redis-standalone 0.0.1) - Standalone Redis Master
```

Just trying to clear up some information on this. Always happy to answer more questions. Obviously we'd love more feedback on helm and https://github.com/deis/charts (which is where we store our default charts/examples for Kubernetes).

/end shameless plug :heart:

**Comment by zmerlynn** [comment]
@bacongobbler: Thanks, I actually corrected in https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-155241457, I just forgot to correct the first comment. I will soon (when I finish putting out another fire). :)

**Comment by technosophos** [comment]
The central problem @zmerlynn raises is _exactly_ what we were trying to address by building Helm. We're not trying to solve deployment, CI/CD, templating, etc. [Helm](helm.sh) is intended to make it easy to find, modify, and use Kubernetes manifests for a wide variety of apps. So I'm really excited to see this post here because it's a topic we are really passionate about.

I would like to weigh in on why we chose a central repository for Helm, because I feel like all of the reasons are relevant to the discussion here:

1. Provide a locus for community interactions
2. Ease of establishing common conventions and practices
3. Fewer maintainers can handle more work
4. Ease of cross-referencing similar issues
5. Better ambiant findability (people discover new charts just by looking around the repo)
6. Less likely to have orphaned projects

In contrast, to check in each manifest in a separate repo will lead to gross fragmentation, lots of maintainership overhead, and a less pleasant user experience.

One of the big advantages of the helm-style repo is that there is nothing there that actually locks you into using Helm. All Helm packages can be managed with kubectl (you just loose the ease of use and the workspace). Helm repos are basically already structured the way you suggest (REAMDE and set of manifests per thing, though we add a [Chart.yaml](https://github.com/deis/helm/blob/master/docs/awsome.md#the-chartyaml-file) metadata file).

**Comment by jackgr** [comment]
@davidopp My suggestion was not that @zmerlynn wait to see what format the community settles on, but rather for /examples to wait on adopting any templating format until we see what the community wants. Clearly, @zmerlynn needs to pick something that helps him deliver turn key solutions, given the project charter.

**Comment by jackgr** [comment]
@technosophos +1 to your points about the motivation for repos that hold multiple types/packages. One repo per type/package would lead to a massive proliferation of repos. 

That said, I believe @zmerlynn is touching on a different issue: whether or not there should be a CNCF owned *official* repository. We agree with him that there should be such a thing, and that CNCF should own it. 

Of course, any such repository needs to be both open, so that anyone can contribute useful content, and well managed, so that it doesn't become a dumping ground. Thankfully, Github already provides the tools we need to solve this dilemma:

A Github-based CNCF-owned repository maintained by a core team would look a lot like Kubernetes:

* The community is free to clone, publish issues and submit pull requests. 
* The maintainers meet with the community discuss submissions, review submissions, accept submissions that make sense in terms of adding value and preserving the integrity of the overall repo, and then merge accepted submissions. 
* CLAs and licences can also follow the Kubernetes model.
* Automated tests validate all submissions, not just individually, but in combination with others (e.g., types using other types in DM).

As @zmerlynn points out, however, an official repo is not mutually exclusive with ad hoc unofficial repos. We've intentionally made it trivially easy to create a repo (e.g., for DM it's just a Github repo with a simple folder structure: /types/type1/v1, /types/type2/v1, types/type2/v2, ...), because we want to encourage individuals, teams, departments, companies, communities, etc., to curate collections of types. Whether they share those collections or use them privately is up to them. There will undoubtedly be unofficial collections shared with the community that gain significant followings.

**Comment by technosophos** [comment]
@timothysc you listed three things that you thought were important:

- well defined semantics around versioning etc.
- ability to earn package acceptance
- feedback mechanics for failed packages

I think you're exactly right about that (we've basically been trying to [formalize](https://github.com/deis/helm/blob/master/docs/architecture.md#the-chart-metadata-file) [that](https://github.com/deis/helm/blob/master/docs/awsome.md#review) [ourselves](https://github.com/deis/helm/issues/96).)

But I'm curious to hear more about what you think about versioning. I'm a huge fan of using a SemVer2 based versioning scheme. I'm not a big fan of having tons of directories inside of a Git repo where each represents a static version. What are your thoughts on how this would look ideally for something like examples.

**Comment by technosophos** [comment]
Having re-read the thread, I feel like consensus seems to be developing around a few central themes:

- There should be a collection of Kubernetes manifests that represent "vetted" packages
- There should be a set of best practices that include versioning semantics, transparency of the package vetting process, and testing
- There has been discussion of whether it should be one package per repo or many packages per repo, but it doesn't seem like anyone wants to limit to _only_ one repo

Is that right? Do we tend to agree on those things?

**Comment by timothysc** [comment]
If I use fedora/EL as a baseline example of how-to create a os/dcos distribution, which is what this is, then an analogous representation would be a list of independent git repositories with defined branches (f21, f22, f23, rawhide) those branches represent the particular distribution in this case it would be k8s11, k8s12, etc.  

Those repositories are then tied to a main repository/db that references repo+branch to create or update a distribution, updates are only allowed after some level of vetting.   So if I did a ~yum update myk8s cluster if would grab the templates from the central location with the transitive closure of the dependency graph for a distribution from the vetted db/repo.  

This allows package federation, with some central control to manage the entire stream.

**Comment by zmerlynn** [comment]
Current bikeshed colors: create a single repo within the https://github.com/kubernetes org named `applications` and start there.

@jackgr, I know from https://github.com/kubernetes/deployment-manager/issues/47 that you're talking about splitting off the type registry. I'm concerned about lack of atomic commit between the DM type registry and something I try to create here. I think they should probably live in the same repo so they can take PRs together (which is also why I'm proposing a joined repo for all packages).

**Comment by timothysc** [comment]
I'd like @smarterclayton opinion here.  The impacts are pretty broad and will shape commercial interest, adoption, and participation.

**Comment by zmerlynn** [comment]
BTW, I'm happy to discuss it at the next community meeting with a slot. I think the next agenda is set but we could maybe devote a chunk. @sarahnovotny? Otherwise it might be past Thanksgiving.

**Comment by jackgr** [comment]
@zmerlynn yes we're thinking about splitting the type registry from the main DM repo, which currently holds both the type registry and the code base. Our thinking is that the type registry doesn't need to live with the code and vice versa. Are you suggesting that we keep them together?

**Comment by zmerlynn** [comment]
@jackgr: I agree that the type registry doesn't need to live with the DM code. I'm saying that I think the type registry should live with the Kubernetes application resources (i.e. the things I'm trying to move out of the `examples/` directory).

**Comment by jackgr** [comment]
After discussing this issue with others offline, it's clear to me that my comments here have been conflating two things:

1. The "official" top level repo for Kubernetes, per the title of this issue.
1. The top level repo for DM.

The former will either have to:

* support multiple tools/formats, in which case it will require discussion and design to finalize, or
* enshrine a single tool/format, in which case it will require quite likely a much longer period of time to let the community gain experience with the various alternatives and express their preferences through adoption.

Either way, it will probably take a while.

The latter is just part of making DM a robust and usable offering. Like other tools, DM needs a top level repo for the reasons outlined in kubernetes/deployment-manager/#82. no point in hobbling DM by delaying its top level repo while we wait for the "official" solution requested by this issue.

Note, also, that creating a top level repo for DM doesn't take anything away from other offerings, most of which already have their own top level repos, and will simply federate with the "official" repo, when it arrives. DM can and should do the same.

The key to unblocking progress on DM is to decouple this "official" repo question from the top level repo for DM. I'd therefore like to ask that all further discussion about the top level repo for DM move to kubernetes/deployment-manager/#82, which is where it belongs.

**Comment by timothysc** [comment]
> enshrine a single tool/format,

never going to happen in open source.  Everyone know that vim is the one true editor amirite ;-)

**Comment by jackgr** [comment]
Well said, @timothysc. It's also unlikely that anyone will want to build an application whose definition spans multiple tool/template formats. Any "official" top level repo will probably need to adopt a single tool/format for the sake of consistency.

And so, this discussion is not likely to lead to the "official" solution any time soon.

Hence, [my comment above](https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-155591386), that:

> For now, since Docker images and Kubernetes YAML files are only de facto standardized foundations at this point in time, /examples should probably limit itself to those technologies.

> Meanwhile, the community can and will, I'm sure, build type registries, package repositories, etc., that people can use for real world applications. Why don't we let that process run for a while, and see which technologies are embraced by the community for that purpose? When facto standards for templatization have emerged, and the community has gained experience with them, then /examples can start to include examples that leverage them.

Going one step further, I'd suggest that we close this issue, let each tool/format use its own top level repo, or any other repo, for that matter, and shift the focus of discussion to the question of applications, per #17447.

**Comment by zmerlynn** [comment]
There's one piece missing if we're going to split, which is where to put Docker images builds that we're hosting (e.g. https://github.com/kubernetes/kubernetes/tree/master/examples/spark/images, etc.). We can just start a top level https://github.com/kubernetes/images/ repo to crack that nut.

**Comment by mattf** [comment]
On 11/18/2015 12:23 PM, Jack Greenfield wrote:
> After discussing this issue with others offline, it's clear to me that
> my comments here have been conflating two things:
>
>  1. The "official" top level repo for Kubernetes, per the title of this
>     issue.
>  2. The top level repo for DM.
>
> The former will either have to:
>
>   * support multiple tools/formats, in which case it will require
>     discussion and design to finalize, or
>   * enshrine a single tool/format, in which case it will require quite
>     likely a much longer period of time to let the community gain
>     experience with the various alternatives and express their
>     preferences through adoption.
>
> Either way, it will probably take a while.
>
> The latter is just part of making DM a robust and usable offering. Like
> other tools, DM needs a top level repo for the reasons outlined in
> kubernetes/deployment-manager/#82
> <https://github.com/kubernetes/kubernetes/pull/82>. no point in hobbling
> DM by delaying its top level repo while we wait for the "official"
> solution requested by this issue.
>
> Note, also, that creating a top level repo for DM doesn't take anything
> away from other offerings, most of which already have their own top
> level repos, and will simply federate with the "official" repo, when it
> arrives. DM can and should do the same.
>
> The key to unblocking progress on DM is to decouple this "official" repo
> question from the top level repo for DM. I'd therefore like to ask that
> all further discussion about the top level repo for DM move to
> kubernetes/deployment-manager/#82
> <https://github.com/kubernetes/kubernetes/pull/82>, which is where it
> belongs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-157785896>.
>

+2

**Comment by zmerlynn** [comment]
*New Current Proposal*

So we arrived at what might be an internal consensus (or at least local stability point) flowing from https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-157785896. Summarizing:

* Create a https://github.com/kubernetes/application-dm-templates repo to host Deployment Manager templates, satisying https://github.com/kubernetes/deployment-manager/issues/82, splitting it from https://github.com/kubernetes/deployment-manager for all the reasons laid out in https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-157785896.

* Create a https://github.com/kubernetes/application-images repo to host image build content for containerized applications: Dockerfiles, startup scripts, config files, etc.

Both of these can be maintained by (separate) teams with fairly open memberships.

If there is interest in an additional repository to host Helm charts, we can create such a repo, but for the upstream https://github.com/helm/charts is the obvious spot. Similarly, if there is interest in hosting OpenShift templates, we can create such a repo, but the upstream https://github.com/openshift/origin/tree/master/examples is the obvious spot. We will link there in the k8s docs as well.

**Comment by davidopp** [comment]
That sounds OK to me.

**Comment by jackgr** [comment]
To be clear, then, the proposed format is application-<tool>-<format>, yes?
So, for example, if we need a repo for helm charts, it would
be application-helm-charts?

On Wednesday, November 18, 2015, David Oppenheimer <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> That sounds OK to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/17027#issuecomment-157925813>
> .
>


-- 
Jack Greenfield | Google

**Comment by zmerlynn** [comment]
@jackgr: Yeah, that was my thought. Mainly because `application-helm-templates` didn't seem correct.

**Comment by timothysc** [comment]
:+1:

**Comment by technosophos** [comment]
That works for the helm team. We've already committed to keeping the `github.com/helm/charts` repo well-maintained and up-to-date. But if there's a separate `application-helm-charts` repo, that's fine by us, too.

**Comment by zmerlynn** [comment]
I see consensus. Let's wait until noon PST tomorrow and call it closed?

**Comment by timothysc** [comment]
perhaps call it closed when the repos exist ;-)

**Comment by zmerlynn** [comment]
Fixed comment to say "discussion closed". :)

**Comment by aronchick** [comment]
+1 (PM weighing in late, after all decisions have been made, as PM is wont to do)

**Comment by zmerlynn** [comment]
Discussion is now closed.

I'm delegating @jackgr to create https://github.com/kubernetes/application-dm-templates, and I'll take care of https://github.com/kubernetes/application-images.

Thank you all for your input!

**Comment by timothysc** [comment]
thx @zmerlynn for taking on this on, and moving the ball forwards!

**Comment by zmerlynn** [comment]
I created a skeleton https://github.com/kubernetes/application-images. I'll push Spark up soon.

**Comment by zmerlynn** [comment]
All of the referenced repos were created and moved done. Closing this issue! :+1:

**Comment by jackgr** [comment]
Nice work, @zmerlynn!
